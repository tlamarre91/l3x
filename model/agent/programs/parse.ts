import { L3xError } from "@/model/errors";
import { Comparison, Instruction, Instructions, isComparison, isInstruction } from "./commands";
import { NamedRegister, NamedRegisters, isNamedRegister } from "./DataDeque";

export class ParseError extends L3xError {
  name = "ParseError";

  constructor(
    public description: string,
    public location: LineAndColumn
  ) {
    const { line, column } = location;
    super(`Parse error at line ${line + 1}, column ${column + 1}: ${description}`);
  }

  withContext(code: string): string {
    // TODO: !! not good. should probably pass code context when error is thrown
    const codeLines = code.split("\n");
    const { line, column } = this.location;
    const codeLine = codeLines[line];
    const pointer = " ".repeat(column) + "^";
    const out = [
      codeLine,
      pointer,
      this.message
    ].join("\n");
    return out;
  }
}

export const Keywords = {
  def: "def",
  ...Instructions,
  ...NamedRegisters
} as const;

export type Keyword = typeof Keywords[keyof typeof Keywords];

export function isKeyword(s: string): s is Keyword {
  return Object.values(Keywords).includes(s as Keyword);
}

export type LineAndColumn = { line: number, column: number };

export interface Token {
  start: LineAndColumn;
  symbol: string;
}

export interface InstructionToken extends Token {
  symbol: Instruction;
}

export interface RefToken extends Token {
  symbol: NamedRegister;
}

export function isRefToken(token: Token | undefined): token is RefToken {
  return isNamedRegister(token?.symbol);
}

export interface ComparisonToken extends Token {
  symbol: Comparison;
}

export function isComparisonToken(token: Token | undefined): token is ComparisonToken {
  return isComparison(token?.symbol);
}

export type StatementType = Instruction | "def";

export function isStatementType(symbol: string): symbol is StatementType {
  return isInstruction(symbol) || symbol === "def";
}

export interface Statement {
  type: StatementType;
  start: LineAndColumn;
  tokens: Token[];
}

export interface TestStatement extends Statement {
  type: "test";
  tokens:
    | [InstructionToken, Token]
    | [InstructionToken, Token, RefToken]
    | [InstructionToken, Token, ComparisonToken, Token]
    | [InstructionToken, Token, ComparisonToken, Token, RefToken];
}

/** Don't use this if you haven't already parsed the statement; it doesn't validate */
export function isTestStatement(statement: Statement): statement is TestStatement {
  return statement.type === "test";
}

export function validateTestStatement(statement: Statement): statement is TestStatement {
  if (statement.tokens[0].symbol !== "test") {
    throw new ParseError(`Expected "test", got ${statement.tokens[0].symbol}`, statement.tokens[0].start);
  }

  if (statement.tokens.length === 2) {
    return true; // TODO: ??? is this enough
  }

  if (statement.tokens.length === 3) {
    const outputToken = statement.tokens[2];

    const valid = isNamedRegister(outputToken.symbol); // TODO: ??? is this enough
    if (!valid) {
      throw new ParseError(`Expected named register, got ${outputToken.symbol}`, outputToken.start);
    }

    return true;
  }

  if (statement.tokens.length === 4 || statement.tokens.length === 5) {
    const comparisonToken = statement.tokens[2];

    const comparisonValid = isComparisonToken(comparisonToken);
    if (!comparisonValid) {
      throw new ParseError(`Expected comparison, got ${comparisonToken.symbol}`, comparisonToken.start);
    }

    const outputValid = statement.tokens.length === 4 || isRefToken(statement.tokens[4]);
    if (!outputValid) {
      throw new ParseError(
        `Expected named register or nothing, got ${statement.tokens[4].symbol}`,
        statement.tokens[4].start
      );
    }

    return true;
  }

  throw new ParseError(`Expected 1, 2, 3, or 4 arguments, got ${statement.tokens.length - 1}`, statement.start);
}

export interface EchoStatement extends Statement {
  type: "echo";
}

export function isEchoStatement(statement: Statement): statement is EchoStatement {
  return statement.type === "echo";
}

export interface GoStatement extends Statement {
  type: "go";
}

export function isGoStatement(statement: Statement): statement is GoStatement {
  return statement.type === "go";
}

export interface MoveStatement extends Statement {
  type: "move";
}

export function isMoveStatement(statement: Statement): statement is MoveStatement {
  return statement.type === "move";
}

export interface SetCursorStatement extends Statement {
  type: "curs";
}

export function isSetCursorStatement(statement: Statement): statement is SetCursorStatement {
  return statement.type === "curs";
}

export interface WriteStatement extends Statement {
  type: "write";
  tokens:
    | [InstructionToken, Token]
    | [InstructionToken, Token, RefToken];
}

/** Don't use this if you haven't already parsed the statement; it doesn't validate */
export function isWriteStatement(statement: Statement): statement is WriteStatement {
  return statement.type === "write";
}

export function validateWriteStatement(statement: Statement): statement is WriteStatement {
  if (statement.tokens.length === 2) {
    return true;
  }

  if (statement.tokens.length !== 3) {
    throw new ParseError(`Expected 1 or 2 arguments, got ${statement.tokens.length - 1}`, statement.start);
  }

  const outputToken = statement.tokens[2];
  if (!isRefToken(outputToken)) {
    throw new ParseError(
      `Expected named register or nothing, got ${outputToken.symbol}`,
      outputToken.start
    );
  }

  return true;
}

export interface Program {
  statements: Statement[];
  codeLines: string[];
}

/** basically just throw if `symbol` starts with "$" but isn't a named register */
export function validateSymbol(symbol: string, start: LineAndColumn): void {
  if (symbol.startsWith("$") && !(isNamedRegister(symbol))) {
    throw new ParseError(`Expected named register, got ${symbol}`, start);
  }
}

export function parseStatement(line: string, lineNumber: number): Statement | undefined {
  // TODO: support any whitespace but make sure to keep character count
  const splitOnSpace = line.split(" ");

  if (splitOnSpace.length === 0) {
    return;
  }

  let currentColumn = 0;
  let lineStartColumn = 0;

  const tokens = new Array<Token>();

  for (const symbol of splitOnSpace) {
    const tokenColumn = currentColumn;
    currentColumn += symbol.length + 1; // add 1 because we split on " "

    if (symbol.length === 0) {
      // handle consecutive whitespace
      lineStartColumn = currentColumn;
      continue;
    }

    const tokenStart: LineAndColumn = {
      line: lineNumber,
      column: tokenColumn
    }

    validateSymbol(symbol, tokenStart);

    if (tokens.length === 0 && !isStatementType(symbol)) {
      throw new ParseError(`First token must be a statement type; got ${symbol}`, tokenStart);
    }

    const token: Token = {
      start: tokenStart,
      symbol
    };

    tokens.push(token);
  }

  if (tokens.length === 0) {
    return;
  }

  const statement = {
    // TODO: for first token, don't append to tokens so we don't store the instruction twice
    type: tokens[0].symbol as StatementType,
    tokens,
    start: {
      line: lineNumber,
      column: lineStartColumn
    },
  } satisfies Statement;

  if (statement.tokens[0].symbol === "test") {
    validateTestStatement(statement);
    statement.type = "test";
  }

  if (statement.tokens[0].symbol === "write") {
    validateWriteStatement(statement);
    statement.type = "write";
  }

  return statement;
}

export function parse(code: string): Program {
  const codeLines = code.split("\n");
  const statements = new Array<Statement>();

  let lineNumber = 0;

  for (const line of codeLines) {
    const statement = parseStatement(line, lineNumber++);

    if (statement == null) {
      continue;
    }

    statements.push(statement);
  }

  return { statements, codeLines };
}
