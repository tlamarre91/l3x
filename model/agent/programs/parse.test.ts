import { expect, it } from "vitest";
import * as programs from ".";
import { ParseError } from "./parse";

it("parse works", () => {
  // (expect this not to compile because of bad namedregister)
  const PARSEABLE_PROGRAM = `def start
echo hello world
test $pd > 0

def loopdeloop
echo again 1
echo again 2
go loopdeloop

def moveover
go f
go start
`;

  const parsed = programs.parse(PARSEABLE_PROGRAM);
  expect(parsed).toMatchSnapshot();
});

it("parse throws on bad syntax", () => {
  const UNPARSEABLE_PROGRAM = `def start
echo hello world
scan file 0 20
branch loopdeloop moveover
`;

  const tryParse = () => programs.parse(UNPARSEABLE_PROGRAM);
  expect(tryParse).toThrowErrorMatchingSnapshot();
});

it("parse error can be printed with context", () => {
  const UNPARSEABLE_PROGRAM = `def start
test $peek > 0
`;

  try {
    programs.parse(UNPARSEABLE_PROGRAM);
    throw Error("Expected ParseError"); // TODO: what's the right way to do this?
  } catch (error: unknown) {
    if (!(error instanceof ParseError)) {
      console.log({ error });
      throw Error("Expected ParseError"); // TODO: what's the right way to do this?
    }

    expect(error.withContext(UNPARSEABLE_PROGRAM)).toMatchSnapshot();
  }
});

it("different parse error can be printed with context", () => {
  const UNPARSEABLE_PROGRAM = `def start
test $pc > 0
test $pc _ 0
`;

  try {
    programs.parse(UNPARSEABLE_PROGRAM);
    throw Error("Expected ParseError"); // TODO: what's the right way to do this?
  } catch (error: unknown) {
    if (!(error instanceof ParseError)) {
      console.log({ error });
      throw Error("Expected ParseError"); // TODO: what's the right way to do this?
    }

    expect(error.withContext(UNPARSEABLE_PROGRAM)).toMatchSnapshot();
  }
});
