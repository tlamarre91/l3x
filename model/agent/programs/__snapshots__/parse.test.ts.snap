// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`different parse error can be printed with context 1`] = `
"test $pc _ 0
         ^
Parse error at line 3, column 10: Expected comparison, got _"
`;

exports[`parse error can be printed with context 1`] = `
"test $peek > 0
     ^
Parse error at line 2, column 6: Expected named register, got $peek"
`;

exports[`parse throws on bad syntax 1`] = `[ParseError: Parse error at line 3, column 1: First token must be a statement type; got scan]`;

exports[`parse works 1`] = `
{
  "codeLines": [
    "def start",
    "echo hello world",
    "test $pd > 0",
    "",
    "def loopdeloop",
    "echo again 1",
    "echo again 2",
    "go loopdeloop",
    "",
    "def moveover",
    "go f",
    "go start",
    "",
  ],
  "statements": [
    {
      "start": {
        "column": 0,
        "line": 0,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 0,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 0,
          },
          "symbol": "start",
        },
      ],
      "type": "def",
    },
    {
      "start": {
        "column": 0,
        "line": 1,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 1,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 1,
          },
          "symbol": "hello",
        },
        {
          "start": {
            "column": 11,
            "line": 1,
          },
          "symbol": "world",
        },
      ],
      "type": "echo",
    },
    {
      "start": {
        "column": 0,
        "line": 2,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 2,
          },
          "symbol": "test",
        },
        {
          "start": {
            "column": 5,
            "line": 2,
          },
          "symbol": "$pd",
        },
        {
          "start": {
            "column": 9,
            "line": 2,
          },
          "symbol": ">",
        },
        {
          "start": {
            "column": 11,
            "line": 2,
          },
          "symbol": "0",
        },
      ],
      "type": "test",
    },
    {
      "start": {
        "column": 0,
        "line": 4,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 4,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 4,
          },
          "symbol": "loopdeloop",
        },
      ],
      "type": "def",
    },
    {
      "start": {
        "column": 0,
        "line": 5,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 5,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 5,
          },
          "symbol": "again",
        },
        {
          "start": {
            "column": 11,
            "line": 5,
          },
          "symbol": "1",
        },
      ],
      "type": "echo",
    },
    {
      "start": {
        "column": 0,
        "line": 6,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 6,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 6,
          },
          "symbol": "again",
        },
        {
          "start": {
            "column": 11,
            "line": 6,
          },
          "symbol": "2",
        },
      ],
      "type": "echo",
    },
    {
      "start": {
        "column": 0,
        "line": 7,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 7,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 7,
          },
          "symbol": "loopdeloop",
        },
      ],
      "type": "go",
    },
    {
      "start": {
        "column": 0,
        "line": 9,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 9,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 9,
          },
          "symbol": "moveover",
        },
      ],
      "type": "def",
    },
    {
      "start": {
        "column": 0,
        "line": 10,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 10,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 10,
          },
          "symbol": "f",
        },
      ],
      "type": "go",
    },
    {
      "start": {
        "column": 0,
        "line": 11,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 11,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 11,
          },
          "symbol": "start",
        },
      ],
      "type": "go",
    },
  ],
}
`;
